src/proposal/plist.mligo

```diff
@@ -39,27 +39,28 @@ let plist_pop (plist_o : proposal_doubly_linked_list option)
 
 [@inline]
 let plist_insert (key, plist_o : proposal_key * (proposal_doubly_linked_list option)): proposal_doubly_linked_list option =
-  match plist_o with
-    | None -> Some
-        { first = key
-        ; last = key
-        ; map = (Big_map.empty : ((proposal_key * plist_direction), proposal_key) big_map)
-        }
-    | Some plist -> Some
-          { plist with
-            last = key
-          ; map =
-              Map.add (plist.last, next) key // Add next link of the old last key
-                (Map.add (key, prev) plist.last plist.map) // Add previous link of the new last key
+  if (plist_mem (key, plist_o) = false) then
+    match plist_o with
+      | None -> Some
+          { first = key
+          ; last = key
+          ; map = (Big_map.empty : ((proposal_key * plist_direction), proposal_key) big_map)
           }
+      | Some plist -> Some
+            { plist with
+              last = key
+            ; map =
+                Map.add (plist.last, next) key // Add next link of the old last key
+                  (Map.add (key, prev) plist.last plist.map) // Add previous link of the new last key
+            }
+  else plist_o // Shouldn't happen, key not existed check should be done before calling this function.
```

Comment:
> When I run these tests locally, it hangs sometimes.

@sras Took me a while to figure this out. This is because the generator sometimes generate the same key that already exists, and inserting that key for some reason cause the hang. Not sure why this happens, I add a check to ensure the inserted key does not exist yet.

This wouldn't happen normally in baseDao, since we have a check to ensure the key is not exist yet. But in our plist contract, we don't have it.

Reply:
Also fixed some other errors, this should be good for re-review.

Reply:
> inserting that key for some reason cause the hang

I think we should find the reson for this happening, and after that may be we can ignore it.

Reply:
@rinn7e Since we are modifying the plist insert by adding a membership check, I think we should make sure that loop is indeed in the ligo code. When I tested, it seems that if I comment the assert statement in the test, the loop disappears. Same happen if I only compare the second element of the tuple. Any chance that loop is in the Haskell counterpart?

If that is the case, we would be adding the membership check for no good reason.

Reply:
That might be because of the laziness. I think you can try to force strict ```!ligoStore``` to force evaluate it which will cause the hang.

Reply:
What I mean is that, it does not hang, because it does not evaluate the ligoStore. So there is no call to the ligo code at all. If there is no call to ligo code, it obviously won't hang since the problem is in the ligo code. Let me know if you don't get what I mean.

Reply:
AFAIU this line https://github.com/tezos-commons/baseDAO/pull/320/files#diff-08120fcaa0dee5bc8d0750d807e29874ae66df02d426d6039ab48a5220126c54R42 

with the membership check was added to fix the hanging test, right?

Reply:
Yeah, this is mainly meant to fix the hanging in the tests.

Reply:
Yes, so if the test hangs because of some flaw in Haskell code, I think the proper way to fix it is by fixing the Haskell code, and not by adding runtime checks on the Ligo code.

Reply:
I couldn't figure the flaw in haskell code. If that is the case, hope to get your suggestion on how to fix it without updating the ligo code.

Reply:
Ok. The issue seems to be in the `plistToList` Haskell function which ends up in a loop when there are duplicate insertions. So the solution would be to detect the loop and cancel the current test. Or better, modify the generator to generate operation lists without duplicate insertions.

Reply:
@sras So I've removed the check in ligo code and add it in the insert call in `callLigo` instead. This should ensure no duplicate key in the list and avoid updating the ligo code.